one/src/ipfs.rs

```diff
@@ -0,0 +1,166 @@
+//! API to create and manage an IPFS service.
+
+use anyhow::{Context, Result};
+use ceramic_kubo_rpc::IpfsService;
+use iroh_p2p::{Config as P2pConfig, DiskStorage, Keychain, Libp2pConfig, Node};
+use iroh_rpc_client::{P2pClient, StoreClient};
+use iroh_rpc_types::{p2p::P2pAddr, store::StoreAddr, Addr};
+use iroh_store::{Config as StoreConfig, Store};
+use std::{path::PathBuf, sync::Arc};
+use tokio::task::{self, JoinHandle};
+use tracing::{error, info};
+
+/// Builder provides an ordered API for constructing an Ipfs service.
+pub struct Builder<S: BuilderState> {
+    state: S,
+}
+
+/// The state of the builder
+pub trait BuilderState {}
+
+/// Initial state of the builder.
+pub struct Init {}
+impl BuilderState for Init {}
+
+/// A builder that has been configured with its store service.
+pub struct WithStore {
+    store: Service<StoreAddr>,
+}
+impl BuilderState for WithStore {}
+
+/// A builder that has been configured with its p2p service.
+pub struct WithP2p {
+    store: Service<StoreAddr>,
+    p2p: Service<P2pAddr>,
+}
+impl BuilderState for WithP2p {}
+
+/// Configure the local block store
+impl Builder<Init> {
+    pub async fn with_store(self, path: PathBuf) -> Result<Builder<WithStore>> {
+        let addr = Addr::new_mem();
+        let config = StoreConfig::with_rpc_addr(path, addr.clone());
+
+        // This is the file RocksDB itself is looking for to determine if the database already
+        // exists or not.  Just knowing the directory exists does not mean the database is
+        // created.
+        let marker = config.path.join("CURRENT");
+
+        let store = if marker.exists() {
+            info!("Opening store at {}", config.path.display());
+            Store::open(config)
+                .await
+                .context("failed to open existing store")?
+        } else {
+            info!("Creating store at {}", config.path.display());
+            Store::create(config)
+                .await
+                .context("failed to create new store")?
+        };
+
+        let rpc_addr = addr.clone();
+        let task =
+            tokio::spawn(async move { iroh_store::rpc::new(rpc_addr, store).await.unwrap() });
```

Comment:
Not a big fan of the unwrap here, maybe split new into a new and run, so we spawn the run?